### PR TITLE
Carry 23584: Passing registry auth token for service create, update, scale

### LINKED ISSUE #23642: breaking protobuf change in swarmkit
There is one final breaking protobuf change from SwarmKit: https://github.com/docker/swarmkit/pull/1000. We wanted to keep the pull options separated clearly from other container options.

This will need to make one of the RCs.

### PR DESCRIPTION
**Pull request by tiborvass**
Closes #23584 (carry)
# 

Currently, it is not possible to create or update services that require pulling of private images. This is a temporary fix to pass an auth token to SwarmKit. The token is retrieved at the client, and passed through as an `x-registry-auth` header which is then retrieved and added to `ContainerSpec` just before making a call to SwarmKit.

This PR updates the following vendored dependencies because they're needed:
- `github.com/docker/engine-api`
- `github.com/docker/swarmkit`
- `github.com/pkg/errors`

Also relevant: https://github.com/docker/swarmkit/pull/965 and https://github.com/docker/swarmkit/pull/1000

Fixes #23642

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @dmcgowan PTAL as well

File: daemon/cluster/executor/container/adapter.go [hunk 1]
```diff
@@ -39,7 +39,7 @@ func newContainerAdapter(b executorpkg.Backend, task *api.Task) (*containerAdapt
 
 func (c *containerAdapter) pullImage(ctx context.Context) error {
 	// if the image needs to be pulled, the auth config will be retrieved and updated
-	encodedAuthConfig := c.container.spec().RegistryAuth
+	encodedAuthConfig := c.container.spec().PullOptions.RegistryAuth
```
Comment (tonistiigi) [inline daemon/cluster/executor/container/adapter.go]: add nil check like in https://github.com/docker/swarmkit/commit/2890e6dde066f46787d33b8fa8684c2f4555b76f#diff-91c312b84ef212e7a7675e103cb76d41R45

File: api/server/router/swarm/cluster_routes.go [hunk 2]
```diff
@@ -107,7 +107,12 @@ func (sr *swarmRouter) createService(ctx context.Context, w http.ResponseWriter,
 		return err
 	}
 
-	id, err := sr.backend.CreateService(service)
+	encodedAuth := ""
+	if auth, ok := r.Header["X-Registry-Auth"]; ok {
```
Comment (thaJeztah) [inline api/server/router/swarm/cluster_routes.go]: Still not sure about this; if another (non-docker CLI) client is sending these headers, which doesn't use this casing

should we add a to-do here for checking case-insensitive?

Comment (tiborvass) [inline api/server/router/swarm/cluster_routes.go]: It seems like the rest of the code is also case sensitive, so the issue is orthogonal to this PR.

File: api/client/registry.go [hunk 3]
```diff
@@ -148,6 +149,34 @@ func (cli *DockerCli) ConfigureAuth(flUser, flPassword, serverAddress string, is
 	return authconfig, nil
 }
 
+// ResolveAuthConfigFromImage retrieves that AuthConfig using the image string
+func (cli *DockerCli) ResolveAuthConfigFromImage(ctx context.Context, image string) (types.AuthConfig, error) {
```
Comment (tonistiigi) [inline api/client/registry.go]: Why public methods?

File: api/client/service/scale.go [hunk 4]
```diff
@@ -60,13 +60,27 @@ func runScale(dockerCli *client.DockerCli, args []string) error {
...
+	headers := map[string][]string{}
 
 	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID)
 
 	if err != nil {
 		return err
 	}
 
+	// TODO(nishanttotla): Is this the best way to get the image?
+	image := service.Spec.TaskTemplate.ContainerSpec.Image
```
Comment (tonistiigi) [inline api/client/service/scale.go]: Does it make sense to send this again on scale/update? This has unexpected behavior then you update service on a client that doesn't have auth info. You don't change anything with the image but it stops working.

Comment (tiborvass) [inline api/client/service/scale.go]: So what do you suggest we do?

Comment (icecrime) [inline api/client/service/scale.go]: Remember you can change the image as part of an update (potentially from a different registry).

Comment (tonistiigi) [inline api/client/service/scale.go]: On update, cli seems to resend this info only if image was updated. That should mean that if something else is updated this info get cleared as I don't see merging with old value in daemon.

Comment (icecrime) [inline api/client/service/scale.go]: Good point...

Comment (tiborvass) [inline api/client/service/scale.go]: Ping @nishanttotla

File: api/client/registry.go (hunk 3 above)
Comment (tonistiigi) [inline api/client/registry.go]: There's one more.

File: api/server/router/swarm/cluster_routes.go (hunk 2 above)
Comment (nishanttotla) [inline api/server/router/swarm/cluster_routes.go]: What @tiborvass said. It doesn't work if this part doesn't use the right capitalization.

File: api/client/service/scale.go (hunk 4 above)
Comment (nishanttotla) [inline api/client/service/scale.go]: A token will always be received, as long as `docker login` was done. The only time the old token gets cleared away is when you logout and try to update/scale a service.

File: api/server/router/swarm/cluster_routes.go (hunk 2 above)
Comment (tiborvass) [inline api/server/router/swarm/cluster_routes.go]: @nishanttotla FYI actually we were wrong and @thaJeztah was right: `r.Header.Get("X-Registry-Auth")` is the way to go.

**Comment by tiborvass** [comment]
Closing since @nishanttotla will implement a new design in his own PR after discussing corner cases.

File: api/server/router/swarm/cluster_routes.go (hunk 2 above)
Comment (nishanttotla) [inline api/server/router/swarm/cluster_routes.go]: This is the right thing, thanks @thaJeztah

**Pull request closed without merging**